docs/code-guide.md

```diff
@@ -0,0 +1,133 @@
+
+Components should be properly broken down into UI elements responsible for a single goal. 
+
+Example folder structure for a hypothetical `UsersList` component:
+
+```
+UsersList/
+├── UsersList.tsx           # Container component
+├── UsersList.test.tsx      
+├── UsersList.stories.tsx      
+├── useUsersList.ts         # Feature-specific hooks
+├── components/             # Presentational components (only imported by container)
+├── └── UsersTable.tsx      # Only create a folder if you have many
+├── └── UsersToolbar.tsx
+├── types.ts                # Only if needed      
+└── utils/                  # Feature-specific utilities
+```
+
+## Coding Patterns
+
+Use the following patterns when designing components:
+- UI components should be a thin wrapper around data, they should handle local state only when necessary.
+- Try to flatten the UI state into a basic calculation, deriving data from props
+- Create a new component abstraction when you're nesting conditional logic or top-level if/else statements. Ternaries are reserved for small, easily readable logic.
+- When complex data manipulation and logic is necessary, make use of custom hooks.
+- Avoid passing whole objects to components when they only need a few properties. It will help clarify which information they rely on.
+- Presentational components should not contain any domain/business logic.
+- Avoid putting dependent logic inside `useEffect`; it causes misdirection of what the logic is doing. Choose to explicitly define logic rather than depend on implicit reactive behavior
+- Prefer state machines over multiple related `useState` calls. Multiple interdependent state variables make code harder to reason about.
+- Avoid `setTimeouts`. They are flaky and usually a _hack_, always provide a comment on _why_ you are using them. This doesn't affect if the "code runs" or not most of the time, but they can introduce subtle bugs that can grow into big issues that aren't obvious until someone goes in and has to spend a lot of time refactoring everything.
+
+## React Hooks Patterns
+
+- Avoid using `useMemo` for processes that are not computationally expensive
+- Functions are memoized with `useCallback` only when necessary:
+  * Functions used as dependencies in `useEffect`, `useMemo`, or other hooks
+  * Functions passed to context providers or returned from custom hooks
+  * Functions passed as props to memoized child components (to prevent unnecessary re-renders)
+- Functions are NOT unnecessarily memoized:
+  * Simple event handlers that aren't passed as props
+  * Functions without dependencies or only used within the same component
+- All hook dependencies (`useEffect`, `useMemo`, `useCallback`) use referentially stable variables
```

Comment:
what is meant by 'use referentially stable variables'?

please explain further

Reply:
It means only variables that are stable across renders, that don't change when their content hasn't.